concepts/dynamic-memory-management/about.md

```diff
@@ -0,0 +1,285 @@
+    // prints The second element of the array is 2
+    printf("The second element of the array is %d\n", numbers[1]);
+    // prints The third element of the array is 3
+    printf("The third element of the array is %d\n", numbers[2]);
+    free(numbers);
+    numbers = NULL;
+}
+```
+
+It's important to remember that the memory newly allocated by `malloc` is not initialized but may hold garbage values.
+Note that the `sizeof` the `numbers` pointer is not the number of the array's elements, nor the total amount of memory allocated.
+The size of the pointer returned by `malloc` is the size of any pointer on that particular system.
+The example code was run on a 64-bit system, so the size of the pointer was 8 byes (multiplied by 8 bits per byte = 64 bits.)
+The pointer type was implicitly converted from pointer-to-void to pointer-to-int.
+It could also have been explicitly converted like so : `int * numbers = (int *) malloc(sizeof(int) * count);`.
+Whether the pointer-to-int points to a single `int` or to an array of `ints` is something the programmer can deal with as seen fit.
+For example
+
+```c
+#include <stdio.h>
+#include <stdlib.h>
+
+int * my_function(int count) {
+    int * numbers = malloc(sizeof(int) * count);
+    return numbers;
+}
+
+int main() {
+    int * numbers = my_function(1);
+    if (!numbers) return -1;
+    // treat numbers as a single value
+    *numbers = 1;
+    // prints numbers is 1
+    printf("numbers is %d\n", *numbers);
+    // treat numbers as an array
+    // prints The first element of the array is 1
+    printf("The first element of the array is %d\n", numbers[0]);
+    numbers[0] = 2;
+    // prints The first element of the array is 2
+    printf("The first element of the array is %d\n", numbers[0]);
+    // treat numbers as a single value again
+    // prints numbers is 2
+    printf("numbers is %d\n", *numbers);    
+    free(numbers);
+    numbers = NULL;    
+}
+```
+If the value(s) will definitely be written to before being read, then to use `malloc` can be an efficient way to allocate memory.
+But, if there is a chance the value(s) will be read before being intialized, then to use `malloc` is risky.
+A safer way to allocate memory is to use `calloc`, which initializes the newly allocated memory to zero.
+
+Note in the examples above that `free(numbers)` was used.
+
+## free
+
+`free` is a function declared in `stdlib.h` which is used for deallocating memory which has been dynamically allocated.
+Dynamically allocated memory stays claimed in memory until released by `free`.
+Some operating systems may automatically free dynamically allocated memory when the program finishes, but it is best practice not to depend on that.
+`free` has one parameter, which is a pointer to the memory which was dynamically allocated.
+The use or lack of use of `free` is associated with several memory issues.
+
+#### memory leak
+
+Not freeing memory after it is no longer needed is referred to as a memory leak.
```

Comment:
```suggestion
Not freeing dynamically allocated memory after the reference to that memory reaches the end of its lifetime, or is otherwise destroyed, is referred to as a memory leak.
```